backend/pkg/k8cache/k8cache.go

```diff
@@ -0,0 +1,308 @@
+// this is the essential part as it gives the clarity about the incoming k8;s requests.
+func UnmarshalCachedata(cacheResource string,
+	cachedData CachedResponseData,
+) (CachedResponseData, error) {
+	err := json.Unmarshal([]byte(cacheResource), &cachedData)
+	if err != nil {
+		return CachedResponseData{}, err
+	}
+
+	return cachedData, nil
+}
+
+// This function is used when serving response from cache to ensure the client
+// receives correct metadata about the response.
+func SetHeader(cacheData CachedResponseData, w http.ResponseWriter) {
+	for idx, header := range cacheData.Headers {
+		w.Header()[idx] = header
+	}
+
+	w.WriteHeader(cacheData.StatusCode)
+}
+
+// MarshallToStore serialize a cacheResponseData struct into JSON []byte.
+// This function is used before storing the K8's response data into cache.
+// ensuring a consistent and structured format for all cached entries.
+func MarshallToStore(cacheData CachedResponseData) ([]byte, error) {
+	jsonByte, err := json.Marshal(cacheData)
+	if err != nil {
+		return nil, err
+	}
+
+	return jsonByte, nil
+}
+
+const gzipEncoding = "gzip"
+
+// This ensures that the cached headers accurately reflect the state of the
+// decompressed body that is being stored, and prevents client side decompression
+// issues serving from cache.
+func SetHeadersTocache(responseHeaders http.Header, encoding string) http.Header {
+	cacheHeader := make(http.Header)
+
+	for idx, header := range responseHeaders {
+		if strings.EqualFold(idx, "Content-Encoding") && encoding == gzipEncoding {
+			continue
+		}
+
+		cacheHeader[idx] = append(cacheHeader[idx], header...)
+	}
+
+	return cacheHeader
+}
+
+// This function checks the user's permission to access the resource.
+// If the user is authorized and has permission to view the resources, it returns true.
+// Otherwise, it returns false if authorization fails.
+func IsAllowed(url *url.URL,
+	kContext *kubeconfig.Context,
+	w http.ResponseWriter,
+	r *http.Request,
+) bool {
+	// Load the right kubeconfig file and context
+	configOverrides := &clientcmd.ConfigOverrides{
+		CurrentContext: kContext.Name,
+	}
+
+	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: kContext.KubeConfigPath},
+		configOverrides,
+	).ClientConfig()
+	if err != nil {
+		return false
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return false
+	}
+
+	apiPath := mux.Vars(r)["api"] // should be e.g. "api/v1/secrets"
+	parts := strings.Split(apiPath, "/")
+
+	last := parts[len(parts)-1]
+
+	review := &authorizationv1.SelfSubjectAccessReview{
+		Spec: authorizationv1.SelfSubjectAccessReviewSpec{
+			ResourceAttributes: &authorizationv1.ResourceAttributes{
+				Resource: last,
+				Verb:     "get",
+			},
+		},
+	}
+
+	result, err := clientset.AuthorizationV1().SelfSubjectAccessReviews().Create(
+		context.TODO(),
+		review,
+		metav1.CreateOptions{},
+	)
+	if err != nil {
+		return false
```

Comment:
If there's an error while checking user access, we shouldn't assume the user is unauthorized. Instead, we should bypass the cache and allow the proxy request to handle the access control. We shouldn't block legitimate users due to a cache related error.